anndata/core/aligned_mapping.py

```diff
@@ -273,44 +298,41 @@ class LayersBase(AlignedMapping):
     attrname = "layers"
     axes = (0, 1)
 
-    # TODO: I thought I had a more elegant solution to overiding this...
-    def copy(self) -> "Layers":
-        d = self._actual_class(self.parent)
-        for k, v in self.items():
-            d[k] = v.copy()
-        return d
-
 
-class Layers(AlignedActualMixin, LayersBase):
-    def __init__(self, parent: "AnnData", vals: Optional[Mapping] = None):
+class _LayersActual(AlignedActual, Layers):
+    def __init__(
+        self, parent: "anndata.AnnData", axis: None = None, **vals,
+    ):
+        assert axis is None
         self._parent = parent
         self._data = dict()
-        if vals is not None:
-            self.update(vals)
+        self.update(vals)
 
 
-class LayersView(AlignedViewMixin, LayersBase):
+class _LayersView(AlignedView, Layers):
     def __init__(
         self,
-        parent_mapping: LayersBase,
-        parent_view: "AnnData",
+        parent_view: "anndata.AnnData",
+        parent_mapping: Layers,
         subset_idx: TwoDIdx,
     ):
         self.parent_mapping = parent_mapping
         self._parent = parent_view
         self.subset_idx = subset_idx
 
 
-LayersBase._view_class = LayersView
-LayersBase._actual_class = Layers
+Layers._View = _LayersView
+Layers._Actual = _LayersActual
 
 
-class PairwiseArraysBase(AlignedMapping):
+class PairwiseArrays(AlignedMapping, ABC):
```

Comment:
It’s like `Path`: Calling `PairwiseArrays()` uses `__new__` to instantiate a subclass directly. There will never be a `Path` or `PairwiseArrays` in existence, but it’s the preferred way to get a subclass instance.

Reply:
`Path` is an abstraction between OSs which gives a consistent interface over different underlying implementations. View and Actual should be functionally different. I think it's a different design pattern for a different use case.

As far as I can tell, none of the AlignedMappings ever use the interface you've described. Currently, I think the `AlignedMapping.__new__` change should be reverted and each class should just use its `__init__`. I don't see a use case, but I do see added indirection.

I think a case which could be useful for this would be to add the `__new__` method to the abstract bases for `Actual` and `View`. Then you could have something like:

```
AlignedMapping(adata, (0,1), vals) -> Layers
AlignedMapping(adata, (0,), vals) -> AxisArrays
AlignedMapping(adata, (0, 0), vals) -> PairwiseArrays
```

But I don't think this needs to happen before a release, and could use a bit more thought.

Reply:
The new interface is used. `Layers(...)`, `AxisArrays(...)` or `PairwiseArrays(...)` goes through this single `__new__`:

```py
from .aligned_mapping import AxisArrays, PairwiseArrays, Layers
...
self._layers = Layers(self, vals=(layers or {}))
# it’s now a Layers._Actual instance
```

I did it this way because we could start by just documenting `AxisArrays`, and in case we add new convenience APIs to e.g. `Layers`, we can also document that, without exposing the private APIs that are used to deal with the view/actual distinction.

> I think a case which could be useful for this would be to add the __new__ method to the abstract bases for `Actual` and `View`.

I think `Layers(adata, vals=...)` is already a nice enough API.

Reply:
> The new interface is used. Layers(...), AxisArrays(...) or PairwiseArrays(...) goes through this single __new__:

But they still all go through their `__init__`? I agree that this code is being used, but it's a change in code with no change to functionality of the API (I don't see changes to where these object are instantiated), which I think should be avoided unless there's and obvious benefit. So far I haven't understood a benefit.

Since this doesn't modify any functionality (at least any functionality I see used), is it something that could happen later?

> I did it this way because we could start by just documenting AxisArrays, and in case we add new convenience APIs to e.g. Layers

How does changing the code for instantiation make this more possible?

Reply:
The difference is that one uses `Layers(...)` instead of `LayersActual(...)` or `LayersView(...)` to construct things, i.e. the `Layers._Actual` and `Layers._View` classes are now implementation details. So when documenting the module, we say:

> There’s the ABC `AlignedMapping` with the following properties: …
> 
> There’s the 3 classes AxisArrays, PairwiseArrays, and Layers which are aligned to the axes …, respectively

If we want to expose it, then we can also add:

> Each class can be `.is_view`, with properties … or `not .is_view` with properties …

I think saying there’s 2×3 classes with a common super class is more instructive than saying there’s 6 classes, 3 of which have the superclass `View` and 2 of which have the superclass `LayersBase`…

Reply:
> The difference is that one uses Layers(...) instead of LayersActual(...) or LayersView(...) to construct things, i.e. the Layers._Actual and Layers._View

Yeah? I don't think that's so different from before, since only `Layers` (which you renamed to `LayersActual`) would ever be constructed outside of the module, and views would be constructed by the `._view` method. I don't any internals here need much more explanation to users than they have now, for layers it's:

> Dictionary-like object with values of the same dimensions as :attr:`X`

To me, that's the only API promise we're making here so we don't need to restructure the classes. I think this whole module is essentially internal, and any docs would be devdocs.

> but I’m tending towards my solution.

I'm not entirely against changing this module. I just think I'm pretty confident it works right now and would like to make a release. I'm not convinced of what changes would be the best, and would rather not have that decision hold a release up. It's also a little hard to see all of the consequences of this change in a PR with so many other changes.

Would you be okay with splitting this out of this PR and we can look at it after a release?